Allow passing fPIC through to OpenSSL compile.

xsacha opened the pull request:
I wasn't able to find any way to get OpenSSL to compile with fPIC.
I tried:
- Using fPIC polly toolchain
- Setting CFLAGS, CXXFLAGS, CMAKE_CXX_FLAGS with fPIC
- Setting CMAKE_POSITION_INDEPENDENT_CODE (just because,.. I know it wouldn't work)

Passing the flag to config manually is the only way I could get this to compile on Linux.

xsacha commented:
configure_opts contains:
-fPIC;threads;shared

ruslo commented on cmake/projects/OpenSSL/schemes/url_sha1_openssl.cmake.in:
```diff
@@ -33,6 +33,11 @@ else()
 endif()
 
+# Pass fPIC through
+if(";${CMAKE_CXX_FLAGS};" MATCHES ";-fPIC;")
```
Why only `-fPIC`? We should pass all flags I think.

xsacha replied:
I do not think it is safe to send all CMAKE_CXX_FLAGS because ./config uses its own inbuilt flags that could conflict, such as -d or -v.
I chose -fPIC because it was causing me issues.

ruslo replied:
> I do not think it is safe to send all CMAKE_CXX_FLAGS because ./config uses its own inbuilt flags that could conflict, such as -d.

If user specify such flags then we should use them. Global flags should came from toolchain and applied to all projects. The issue you have reported is exactly about it. Also it make sense to pass CMAKE_CXX_FLAGS_RELEASE and think about multiple-config installation of OpenSSL :)

> I chose -fPIC because it was causing me issues

I see nothing special about `-fPIC`, I expect same issue with `-fsanize=*` flags for instance.

xsacha replied:
Also, OpenSSL is a C project. Should I use CMAKE_C_FLAGS instead?

ruslo replied:
Yes, I guess so.

ruslo replied:
I see that you have pushed new commit. If you're done, let me know about it because I have no idea that you've finished updates or it's just pushed a work-in-progress stuff:
* https://github.com/ruslo/hunter/blob/master/.github/CONTRIBUTING.rst
* https://github.com/ruslo/hunter/wiki/dev.contribution

xsacha replied:
Was just waiting for build to complete on travis.
Yes, it's complete as long as you like that change.

ruslo replied:
> Was just waiting for build to complete on travis

Basic Travis testing will only run unit tests. There will be no OpenSSL tests triggered.

xsacha replied:
Ok, thanks.
Updated again.

ruslo commented:
Please update `PACKAGE_INTERNAL_DEPS_ID`:
* https://github.com/ruslo/hunter/blob/4d4f792f90111e4df09f98b7d7d81e0396eee40b/cmake/projects/OpenSSL/hunter.cmake#L266

We need to do it so cache system will know that package changed even if there is no modification of sources. See details: https://github.com/ruslo/hunter/wiki/dev.modules#hunter_download

ruslo commented:
> Updated again

Thanks, testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1022
* https://travis-ci.org/ingenue/hunter/builds/163318873

ruslo commented:
Some tests failed:
* https://travis-ci.org/ingenue/hunter/jobs/163318880

```
gcc -I. -I.. -I../include  -DOPENSSL_THREADS -D_REENTRANT -DDSO_DLFCN -DHAVE_DLFCN_H -g -fsanitize=address -Wa,--noexecstack -m64 -DL_ENDIAN -O3 -Wall -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DRC4_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DWHIRLPOOL_ASM -DGHASH_ASM -DECP_NISTZ256_ASM   -c -o cryptlib.o cryptlib.c

cc1: error: unrecognized command line option '-fsanitize=address'
```

This is because OpenSSL use `gcc` instead of `clang` from toolchain. I guess we should pass `CMAKE_C_COMPILER` too.

xsacha commented:
Updated again to pass CMAKE_C_COMPILER (hopefully)

ruslo commented on cmake/projects/OpenSSL/schemes/url_sha1_openssl.cmake.in:
```diff
@@ -30,9 +30,12 @@ if(APPLE)
   set(configure_command "./Configure")
   set(configure_opts "darwin64-x86_64-cc")
```
I think this branch should be updated too. Probably it make sense to add `CC=*` out of the `if` conditions (for both APPLE and non-APPLE).

xsacha replied:
Ok, updated to do it for APPLE as well.

ruslo commented:
> Ok, updated to do it for APPLE as well.

Thanks, testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1025
* https://travis-ci.org/ingenue/hunter/builds/163637622

ruslo commented:
> testing

Not working (https://travis-ci.org/ingenue/hunter/jobs/163637623):
```
/bin/sh: 1: CC=/usr/bin/gcc ./config: not found
```

xsacha commented:
Oops, updated again :)
Checked it builds locally this time.

ruslo commented:
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1030
* https://travis-ci.org/ingenue/hunter/builds/163974409

ruslo commented:
`analyze` toolchain is not working:
```
/home/travis/build/ingenue/hunter/polly-master/scripts/clang-analyze.sh: line 20: clang: command not found
```

This toolchain set script `clang-analyze.sh` as a compiler and internally invoke `clang --analyze`. For some reason `clang` compiler can't be found from inside OpenSSL build script. Feel free to investigate. However I'm okay with excluding `analyze` tests since they was not working correctly anyway. In this case please send patch to https://github.com/ingenue/hunter/tree/pkg.openssl branch for excluding toolchain and add a comment with the link to CI jobs.

xsacha commented:
I updated ingenue pkg.openssl here: https://github.com/ingenue/hunter/pull/31
Also updated Android build error (gcc complains about -m64 being sent to it by OpenSSL)

ruslo commented:
> Also updated Android build error (gcc complains about -m64 being sent to it by OpenSSL)

Don't see this change. You have it locally? Does it mean that OpenSSL for Android works now?

xsacha commented:
I mean: prior to this change, Android build would actually do a host build. Now, it builds Android but fails because "arm-gnueabihf-gcc: unrecognized option 'm64'"

ruslo commented:
Okay, I see.

xsacha commented:
I have rebased the change in pkg.openssl without the pr534 changes

ruslo commented:
Testing:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1068
* https://travis-ci.org/ingenue/hunter/builds/165189443

ruslo commented:
Uploading:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.1071
* https://travis-ci.org/ingenue/hunter/builds/165211372

ruslo commented:
> Uploading

Done.

Released: https://github.com/ruslo/hunter/releases/tag/v0.16.7
